### ISSUE TITLE
Show menu announces an extra VoiceOver item

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#56850]
b/154075528

In iOS, the usage of PopupMenuItem and some basic iOS overlay introduces an extra VoiceOver announcement for the Semantic label of the underlying page.

```dart
import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Semantics(
        namesRoute: true,
        scopesRoute: true,
        explicitChildNodes: true,
        label: 'My awesome test!',
        child: MyHomePage(title: 'Flutter Demo Home Page'),
      ),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    final button = FlatButton(
      onPressed: () async {
        ImagePicker.pickImage(source: ImageSource.gallery);
      },
      child: Text('Pick Image'),
    );
    return Scaffold(
      appBar: AppBar(
          title: Text(widget.title),
          actions: [
            PopupMenuButton<FlatButton>(
              onSelected: (value) {
                value.onPressed();
              },
              itemBuilder: (context) {
                return [
                  PopupMenuItem(
                    value: button,
                    child: button.child,
                  )
                ];
              },
            ),
          ]),
      body: Center(
        child: Text(
          'Button tapped $_counter time${_counter == 1 ? '' : 's'}.',
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```

I used the image picker plugin here and replaced the view controller it launches with this:

```objc
  if ([@"pickImage" isEqualToString:call.method]) {
    UIViewController *controller = [[UIViewController alloc] initWithNibName:nil bundle:nil];
    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:controller];
    navController.navigationBarHidden = NO;
    [_viewController presentViewController:navController animated:YES completion:nil];
```

Steps:

1. Launch App.
2. Turn on VoiceOver.
3. Navigate and open the overflow menu in the title.
4. Tap on Pick Image.
5. Notice that when the view controller opens, app announces: "My awesome test!".

Expected behavior:
It should not announce My awesome test.

**Comment by HansMuller** [comment]
CC @goderbauer

**Comment by goderbauer** [comment]
I can reproduce this bug and looking at our implementation also explain why it is happening:

In the iOS accessibility bridge we always announce when the topmost Flutter route changes. In this interaction from Flutter's perspective you're first on the Home Route (with title "My awesome test!"), then on the Dropdown-Menu Route (no title), and then you're back on the Home Route, which causes the title to get announced again. Flutter doesn't know that it's covered by a native view (the image picker). From its perspective, the Home Route is the topmost route and it's title gets announced.

The announcement happens here in the engine:

https://github.com/flutter/engine/blob/d3bde19c1519a8f388cc08ee1a6bc054e9406d39/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm#L141

Maybe we can add a check and skip the route announcement when Flutter is covered by a native view (as it is here with the image picker). /cc @gaaclarke @xster

**Comment by goderbauer** [comment]
To the customer, the comment made in https://github.com/flutter/flutter/issues/55306#issuecomment-620916841 also applies here:

> On iOS it actually seems to be fairly uncommon to actually name a route. If you remove the namesRoute flag (maybe conditionally only when running on iOS), this problem goes away.

**Comment by cbracken** [comment]
Unassigning as it's unlikely the engine team currently has time to focus on this.

**Comment by xster** [comment]
@goderbauer did you mean just specifically changing the `routeChange` path of the announcement or disabling Flutter voiceover bridging altogether when the FlutterViewController is being navigated away from? I suspect the latter is the right approach but maybe you had something in mind.

**Comment by xster** [comment]
@gaaclarke can I send this customer bug to you?

**Comment by gaaclarke** [comment]
@xster @goderbauer Are you just talking about adding this?
```diff
--- a/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm
+++ b/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm
@@ -136,7 +136,7 @@ void AccessibilityBridge::UpdateSemantics(flutter::SemanticsNodeUpdates nodes,
   [objects_ removeObjectsForKeys:doomed_uids];
 
   layoutChanged = layoutChanged || [doomed_uids count] > 0;
-  if (routeChanged) {
+  if (routeChanged && TopmostViewControllerIsFlutter()) {
     NSString* routeName = [lastAdded routeName];
     UIAccessibilityPostNotification(UIAccessibilityScreenChangedNotification, routeName);
   } else if (layoutChanged) {
```
I can do that easily.  I'm not familiar with the route change announcement though so I'm not sure of the ramifications.

**Comment by gaaclarke** [comment]
I ran through this with the debugger.  The modal view controller is a bit of a red herring.  The route is announced because when the main view is visible `[lastAdded routeName]` is "My awesome test", then when the popup menu is presented, it is `nil`.  Then when the popup menu is dismissed it is "My awesome test" again before the modal view controller is even presented.  It's the dismissal of the popup menu that triggers the announcement, not the presentation of the modal view.  The modal view doesn't even exist yet at that point.

**Comment by gaaclarke** [comment]
@goderbauer it seems like maybe we shouldn't be sending a routeChange node when the popup menu is shown right?  Will it ever have a name other than `nil`?

**Comment by gaaclarke** [comment]
Did some digging and it seems appropriate to do a route change for a popup menu since it is a modal barrier ( https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/widgets/navigator.dart#L1127 ).

**Comment by gaaclarke** [comment]
Further investigation shows that the UIImagePickerViewController does in fact exist and has been presented before the route change is posted.  The problem is that since the picker view controller is getting presented with an animation `[[[[UIApplication sharedApplication] keyWindow] rootViewController] presentedViewController] == nil` until the animation is complete.  That is why attempts to introspect on it failed.

**Comment by gaaclarke** [comment]
I printed out the following at the point where we wanted to send the screen change notification during the presentation of the modal view controller:
```objc
NSLog(@"presentedViewController %@", [[[[UIApplication sharedApplication] keyWindow] rootViewController] presentedViewController]);
NSLog(@"presentationController %@", [[[[UIApplication sharedApplication] keyWindow] rootViewController] presentationController]);
NSLog(@"presentingViewController %@", [[[[[UIApplication sharedApplication] keyWindow] rootViewController] presentationController] presentingViewController]);
NSLog(@"presentedViewController %@", [[[[[UIApplication sharedApplication] keyWindow] rootViewController] presentationController] presentedViewController]);
NSLog(@"transitionCoordinator %@", [[[[UIApplication sharedApplication] keyWindow] rootViewController] transitionCoordinator]);
NSLog(@"popoverPresentationController %@", [[[[UIApplication sharedApplication] keyWindow] rootViewController] popoverPresentationController]);
```
results:
```
2020-05-21 16:36:39.130527-0700 Runner[14894:374809] presentedViewController (null)
2020-05-21 16:36:39.130735-0700 Runner[14894:374809] presentationController <_UIPageSheetPresentationController: 0x10a08d700>
2020-05-21 16:36:39.130779-0700 Runner[14894:374809] presentingViewController (null)
2020-05-21 16:36:39.130865-0700 Runner[14894:374809] presentedViewController <FlutterViewController: 0x109833000>
2020-05-21 16:36:39.130910-0700 Runner[14894:374809] transitionCoordinator (null)
2020-05-21 16:36:39.130952-0700 Runner[14894:374809] popoverPresentationController (null)
```

This means there doesn't appear to be a way to query if we are transitioning to a modal view controller.  We can customize `FlutterViewController` to send out notifications surrounding presenting view controllers and use that context to determine if we should suppress the notification.

**Comment by gaaclarke** [comment]
I repurposed https://github.com/flutter/engine/pull/18544 to include a fix for the issue. Now we ignore route change notifications if we are presenting a modal view controller on the associated FlutterViewController.  This gives us the desired result.  One weird thing is that no route change is announced when the modal view controller is dismissed though.  If we care about that we should file another issue, that was always the case before and after this PR though.

**Comment by goderbauer** [comment]
> I repurposed flutter/engine#18544 to include a fix for the issue. Now we ignore route change notifications if we are presenting a modal view controller on the associated FlutterViewController. This gives us the desired result. One weird thing is that no route change is announced when the modal view controller is dismissed though. If we care about that we should file another issue, that was always the case before and after this PR though.

This all sounds good to me and is what I expected as a solution for this issue.
